### PR TITLE
[ Refactor ] Support connect method Ack handler.

### PR DESCRIPTION
**Pull request by mattisonchao**
## Motivation

To clean the code, better support the many features of MQTT5.

## Modification

base on #324 and #325 , 

I design a graceful ``AckHandler`` API to handle all ack method.

- Change connection init method.
- Use delegate to wrapper ack handler.
- Use Netty ``MqttMessageBuilder#conAck`` to instead of   ``MqttConnAckMessageHelper``.

### CODE REVIEW

**Comment by mattisonchao** [comment]
@Technoboy-  PTAL :)

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/Connection.java [hunk 1]
```diff
@@ -117,8 +119,8 @@ private void addIdleStateHandler() {
                 Math.round(keepAliveTime * 1.5f)));
     }
 
-    public void sendConnAck() {
-        ackHandler.sendConnAck(this);
+    public AckHandlerDelegate ack() {
```
Comment (Technoboy-) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/Connection.java:122]: ack() -> ackHandler()

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AbstractAckHandler.java [hunk 2]
```diff
@@ -59,4 +51,33 @@ public void sendConnAck(Connection connection) {
             }
         });
     }
+
+    abstract MqttMessage getConnAckQosNotSupportedMessage(Connection connection);
```
Comment (Technoboy-) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AbstractAckHandler.java:55]: getConnAckQosNotSupportedMessage -> getQosNotSupportedMessage

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AbstractAckHandler.java [hunk 3]
```diff
@@ -59,4 +51,33 @@ public void sendConnAck(Connection connection) {
...
+    @Override
+    public void sendConnClientIdentifierInvalidAck(Connection connection) {
+        log.error("The MQTT client ID cannot be empty. the remote address is {}",
+                connection.getChannel().remoteAddress());
+        MqttMessage connAckClientIdentifierInvalidMessage =
+                getConnAckClientIdentifierInvalidMessage(connection);
+        connection.sendThenClose(connAckClientIdentifierInvalidMessage);
+    }
+
+    abstract MqttMessage getConnAuthenticationFailAck(Connection connection);
```
Comment (Technoboy-) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AbstractAckHandler.java:75]: getConnAuthenticationFailAck -> getAuthenticationFailMessage

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AbstractAckHandler.java [hunk 4]
```diff
@@ -59,4 +51,33 @@ public void sendConnAck(Connection connection) {
...
+    abstract MqttMessage getConnAckQosNotSupportedMessage(Connection connection);
+
+    @Override
+    public void sendConnNotSupportedAck(Connection connection) {
+        log.error("MQTT protocol qos not supported. CId={}", connection.getClientId());
+        MqttMessage connAckQosNotSupportedMessage = getConnAckQosNotSupportedMessage(connection);
+        connection.sendThenClose(connAckQosNotSupportedMessage);
+    }
+
+    abstract MqttMessage getConnAckClientIdentifierInvalidMessage(Connection connection);
```
Comment (Technoboy-) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AbstractAckHandler.java:64]: getConnAckClientIdentifierInvalidMessage -> getClientIdentifierInvalidMessage

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AckHandlerDelegate.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    private AckHandlerDelegate(Connection connection, AckHandler delegate) {
+        this.connection = connection;
+        this.delegate = delegate;
+    }
+
+    public static AckHandlerDelegate delegate(Connection connection, AckHandler handler) {
+        return new AckHandlerDelegate(connection, handler);
+    }
+
+    public void connAck() {
```
Comment (Technoboy-) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AckHandlerDelegate.java:31]: connAck -> sendConnAck

**Comment by mattisonchao** [comment]
@Technoboy-  PTAL

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AckHandler.java [hunk 6]
```diff
@@ -13,12 +13,20 @@
...
 
+import io.netty.channel.Channel;
 import io.streamnative.pulsar.handlers.mqtt.Connection;
 
 /**
  * Ack handler.
  */
 public interface AckHandler {
 
+    void sendConnNotSupported(Channel channel);
```
Comment (Technoboy-) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AckHandler.java:24]: sendConnNotSupported -> sendConnVersionNotSupported

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AckHandler.java [hunk 7]
```diff
@@ -13,12 +13,20 @@
...
+
+    void sendConnClientIdentifierInvalid(Channel channel);
+
+    void sendConnAuthenticationFail(Channel channel);
```
Comment (Technoboy-) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AckHandler.java:28]: Why change the above method from Connection to channel ?

**Pull request closed without merging**